### PR TITLE
Return HTTP 405 'Method Not Allowed' when appropriate

### LINKED ISSUE #2: Shouldn't faguard return 405 method not allowed instead of 404 when url exists?
Thank you for *Falguard*, it is very useful.

I have a question: Say I have a resource published on `/hello` that only implement `on_get`. 

If I try to access it with a POST request, *Falguard* returns a 404 error, whereas it should be a 405, shouldn't it ?

### PR DESCRIPTION
**Pull request by rtxm**
if URI exists in spec but requested method is wrong, we should return a 405 error with the list of allowed methods if any.

Fixes #2

### CODE REVIEW

**Comment by rtxm** [comment]
Ready for review.

File: tests/test_error.py [hunk 1]
```diff
@@ -75,4 +75,9 @@ def test_middleware_validation_error(
 
 
 def test_no_operation_found(middleware_app):
-    middleware_app.put('/tests', status=404)
+    middleware_app.put('/missing_tests', status=404)
+
+
+def test_method_not_allowed(middleware_app):
+    middleware_app.put('/tests', status=405)
```
Comment (gwaramadze) [inline tests/test_error.py]: It should also test `Allow` header.
```suggestion
    response = middleware_app.put('/tests', status=405)
    assert response.headers['Allow'] == 'GET, POST'
```

File: falguard.py [hunk 2]
```diff
@@ -69,13 +69,28 @@ def __init__(self, spec_path):
...
+        allowed_methods = [method for method in falcon.HTTP_METHODS
+                           if self._spec.get_op_for_request(method, uri)]
+
+        return (falcon.HTTPMethodNotAllowed(allowed_methods) if allowed_methods
+                else falcon.HTTPNotFound)
+
     def process_resource(self, request, response, resource, params):
         bravado_request = _BravadoRequest(request, params)
 
+        uri = request.uri_template
```
Comment (gwaramadze) [inline falguard.py:76]: Hmm, I will mix English and Python for this comment. Looks like you can `raise falcon.HTTPNotFound` at this point `if not uri`.

```suggestion
        uri = request.uri_template
        if not uri:
            raise falcon.HTTPNotFound
```

File: tests/test_error.py [hunk 3]
```diff
@@ -75,4 +75,9 @@ def test_middleware_validation_error(
...
+
```
Comment (gwaramadze) [inline tests/test_error.py:84]: Looks like there is one blank line too many at the end of this module. I will need to update Travis to run `py.test ... --pep8 --flakes` or something similar

File: falguard.py [hunk 4]
```diff
@@ -69,13 +69,28 @@ def __init__(self, spec_path):
...
-            request.method.upper(), request.uri_template)
+            request.method.upper(), uri)
         if not operation:
-            raise falcon.HTTPNotFound
+            if uri:
+                # The URI exists but the method is wrong.
+                # So we are going to reply with an error 405.
+                raise self._build_http_not_allowed_error(uri)
+            else:
+                raise falcon.HTTPNotFound
```
Comment (gwaramadze) [inline falguard.py]: As mentioned in one of the previous comments, you can move this logic right after line 84. What I didn't mention is that this path is not covered by tests. I will need to update Travis to run `py.test ... --cov falguard --cov-fail-under 100`

File: falguard.py [hunk 5]
```diff
@@ -69,13 +69,28 @@ def __init__(self, spec_path):
     def __call__(self, request, response, resource, params):
         self.process_resource(request, response, resource, params)
 
+    def _build_http_not_allowed_error(self, uri):
+        # Error 405 requires we provide the list of allowed methods
+        # for this URI. If None is found, then we produce a 404 error.
+        allowed_methods = [method for method in falcon.HTTP_METHODS
+                           if self._spec.get_op_for_request(method, uri)]
```
Comment (gwaramadze) [inline falguard.py]: This way get_op_for_request will run twice for the original method.

File: falguard.py [hunk 6]
```diff
@@ -69,13 +69,28 @@ def __init__(self, spec_path):
...
+
+        return (falcon.HTTPMethodNotAllowed(allowed_methods) if allowed_methods
+                else falcon.HTTPNotFound)
```
Comment (gwaramadze) [inline falguard.py]: This whole private method looks a bit unnecessary. My suggestion for refactoring right in the process_resource method:

```python
        uri = request.uri_template
        if not uri:
            raise falcon.HTTPNotFound

        method = request.method.upper()
        get_op = functools.partial(
            self._spec.get_op_for_request,
            path_pattern=uri,
        )

        operation = get_op(method)
        if not operation:
            allowed = [
                other_method for other_method in falcon.HTTP_METHODS
                if other_method != method and get_op(other_method)
            ]
            if allowed:
                raise falcon.HTTPMethodNotAllowed(allowed)
            raise falcon.HTTPNotFound
```

**Review by gwaramadze — CHANGES_REQUESTED** [review]
Hello, thank you for your time and I am very very sorry for such a late review. I am requesting a couple of changes. But in general I like the idea and it is a useful step. Please also after you're done squash your commits into one as currently this change looks pretty atomic.

File: tests/test_error.py (hunk 3 above)
Comment (rtxm) [reply tests/test_error.py:84]: This one is required by PEP8 indeed (line 82 must end with carriage return)

File: falguard.py (hunk 5 above)
Comment (rtxm) [reply falguard.py]: Well, I was wondering, which is cheaper:
- one more call to `get_op_for_request` (that does a dict key lookup inside) ;
- as many comparisons as they are elements in `HTTP_METHODS`?

File: falguard.py (hunk 2 above)
Comment (rtxm) [reply falguard.py:76]: In fact the whole test in unnecessary since the `process_resource` method is called after routing is done. I removed it.

**Comment by rtxm** [comment]
I Made the changes (except one still open to discussion above) and squashed the commits.

File: falguard.py (hunk 5 above)
Comment (gwaramadze) [reply falguard.py]: It's ok. Readability first.

**Pull request merged**